File: BleModule/src/main/java/com/topdon/ble/ConnectionConfiguration.java
package com.topdon.ble;

import android.bluetooth.BluetoothDevice;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.core.util.Pair;

import java.util.*;

public class ConnectionConfiguration {

    public static final int TRY_RECONNECT_TIMES_INFINITE = -1;
    @NonNull
    final List<Pair<Integer, Integer>> scanIntervalPairsInAutoReconnection;
    private final Map<String, WriteOptions> defaultWriteOptionsMap = new HashMap<>();
    int discoverServicesDelayMillis = 600;
    int connectTimeoutMillis = 10000;
    int requestTimeoutMillis = 3000;
    int tryReconnectMaxTimes = TRY_RECONNECT_TIMES_INFINITE;
    int reconnectImmediatelyMaxTimes = 3;
    boolean isAutoReconnect = true;
    @RequiresApi(Build.VERSION_CODES.M)
    int transport = BluetoothDevice.TRANSPORT_LE;
    @RequiresApi(Build.VERSION_CODES.O)
    int phy = BluetoothDevice.PHY_LE_1M_MASK;

    public ConnectionConfiguration() {
        scanIntervalPairsInAutoReconnection = new ArrayList<>();
        scanIntervalPairsInAutoReconnection.add(Pair.create(0, 2000));
        scanIntervalPairsInAutoReconnection.add(Pair.create(1, 5000));
        scanIntervalPairsInAutoReconnection.add(Pair.create(3, 10000));
        scanIntervalPairsInAutoReconnection.add(Pair.create(5, 30000));
        scanIntervalPairsInAutoReconnection.add(Pair.create(10, 60000));
    }

    public ConnectionConfiguration setDiscoverServicesDelayMillis(int discoverServicesDelayMillis) {
        this.discoverServicesDelayMillis = discoverServicesDelayMillis;
        return this;
    }

    public ConnectionConfiguration setConnectTimeoutMillis(int connectTimeoutMillis) {
        if (requestTimeoutMillis >= 1000) {
            this.connectTimeoutMillis = connectTimeoutMillis;
        }
        return this;
    }

    public ConnectionConfiguration setRequestTimeoutMillis(int requestTimeoutMillis) {
        if (requestTimeoutMillis >= 1000) {
            this.requestTimeoutMillis = requestTimeoutMillis;
        }
        return this;
    }

    public ConnectionConfiguration setTryReconnectMaxTimes(int tryReconnectMaxTimes) {
        this.tryReconnectMaxTimes = tryReconnectMaxTimes;
        return this;
    }

    public ConnectionConfiguration setReconnectImmediatelyMaxTimes(int reconnectImmediatelyMaxTimes) {
        this.reconnectImmediatelyMaxTimes = reconnectImmediatelyMaxTimes;
        return this;
    }

    public ConnectionConfiguration setAutoReconnect(boolean autoReconnect) {
        isAutoReconnect = autoReconnect;
        return this;
    }

    @RequiresApi(Build.VERSION_CODES.M)
    public ConnectionConfiguration setTransport(int transport) {
        this.transport = transport;
        return this;
    }

    @RequiresApi(Build.VERSION_CODES.O)
    public ConnectionConfiguration setPhy(int phy) {
        this.phy = phy;
        return this;
    }

    public ConnectionConfiguration setScanIntervalPairsInAutoReconnection(List<Pair<Integer, Integer>> parameters) {
        Inspector.requireNonNull(parameters, "parameters can't be null");
        scanIntervalPairsInAutoReconnection.clear();
        scanIntervalPairsInAutoReconnection.addAll(parameters);
        return this;
    }

    public ConnectionConfiguration setDefaultWriteOptions(UUID service, UUID characteristic, WriteOptions options) {
        Inspector.requireNonNull(service, "service can't be null");
        Inspector.requireNonNull(characteristic, "characteristic can't be null");
        Inspector.requireNonNull(options, "options can't be null");
        defaultWriteOptionsMap.put(service + ":" + characteristic, options);
        return this;
    }

    @Nullable
    WriteOptions getDefaultWriteOptions(UUID service, UUID characteristic) {
        return defaultWriteOptionsMap.get(service + ":" + characteristic);
    }
}

File: ble-core/src/main/java/com/mpdc4gsr/ble/core/ConnectionConfiguration.kt
package com.mpdc4gsr.ble.core

import android.bluetooth.BluetoothDevice
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.util.Pair
import java.util.UUID

class ConnectionConfiguration {
    val scanIntervalPairsInAutoReconnection: MutableList<Pair<Int?, Int?>?>
    private val defaultWriteOptionsMap: MutableMap<String?, WriteOptions?> = HashMap<String?, WriteOptions?>()
    var discoverServicesDelayMillis: Int = 600
    var connectTimeoutMillis: Int = 10000
    var requestTimeoutMillis: Int = 3000
    var tryReconnectMaxTimes: Int = TRY_RECONNECT_TIMES_INFINITE
    var reconnectImmediatelyMaxTimes: Int = 3
    var isAutoReconnect: Boolean = true

    @RequiresApi(Build.VERSION_CODES.M)
    var transport: Int = BluetoothDevice.TRANSPORT_LE

    @RequiresApi(Build.VERSION_CODES.O)
    var phy: Int = BluetoothDevice.PHY_LE_1M_MASK

    init {
        scanIntervalPairsInAutoReconnection = ArrayList<Pair<Int?, Int?>?>()
        scanIntervalPairsInAutoReconnection.add(Pair.create<Int?, Int?>(0, 2000))
        scanIntervalPairsInAutoReconnection.add(Pair.create<Int?, Int?>(1, 5000))
        scanIntervalPairsInAutoReconnection.add(Pair.create<Int?, Int?>(3, 10000))
        scanIntervalPairsInAutoReconnection.add(Pair.create<Int?, Int?>(5, 30000))
        scanIntervalPairsInAutoReconnection.add(Pair.create<Int?, Int?>(10, 60000))
    }

    fun setDiscoverServicesDelayMillis(discoverServicesDelayMillis: Int): ConnectionConfiguration {
        this.discoverServicesDelayMillis = discoverServicesDelayMillis
        return this
    }

    fun setConnectTimeoutMillis(connectTimeoutMillis: Int): ConnectionConfiguration {
        if (requestTimeoutMillis >= 1000) {
            this.connectTimeoutMillis = connectTimeoutMillis
        }
        return this
    }

    fun setRequestTimeoutMillis(requestTimeoutMillis: Int): ConnectionConfiguration {
        if (requestTimeoutMillis >= 1000) {
            this.requestTimeoutMillis = requestTimeoutMillis
        }
        return this
    }

    fun setTryReconnectMaxTimes(tryReconnectMaxTimes: Int): ConnectionConfiguration {
        this.tryReconnectMaxTimes = tryReconnectMaxTimes
        return this
    }

    fun setReconnectImmediatelyMaxTimes(reconnectImmediatelyMaxTimes: Int): ConnectionConfiguration {
        this.reconnectImmediatelyMaxTimes = reconnectImmediatelyMaxTimes
        return this
    }

    fun setAutoReconnect(autoReconnect: Boolean): ConnectionConfiguration {
        isAutoReconnect = autoReconnect
        return this
    }

    @RequiresApi(Build.VERSION_CODES.M)
    fun setTransport(transport: Int): ConnectionConfiguration {
        this.transport = transport
        return this
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun setPhy(phy: Int): ConnectionConfiguration {
        this.phy = phy
        return this
    }

    fun setScanIntervalPairsInAutoReconnection(parameters: MutableList<Pair<Int?, Int?>?>?): ConnectionConfiguration {
        Inspector.requireNonNull<MutableList<Pair<Int?, Int?>?>?>(parameters, "parameters can't be null")
        scanIntervalPairsInAutoReconnection.clear()
        scanIntervalPairsInAutoReconnection.addAll(parameters!!)
        return this
    }

    fun setDefaultWriteOptions(service: UUID?, characteristic: UUID?, options: WriteOptions?): ConnectionConfiguration {
        Inspector.requireNonNull<UUID?>(service, "service can't be null")
        Inspector.requireNonNull<UUID?>(characteristic, "characteristic can't be null")
        Inspector.requireNonNull<WriteOptions?>(options, "options can't be null")
        defaultWriteOptionsMap.put(service.toString() + ":" + characteristic, options)
        return this
    }

    fun getDefaultWriteOptions(service: UUID?, characteristic: UUID?): WriteOptions? {
        return defaultWriteOptionsMap.get(service.toString() + ":" + characteristic)
    }

    companion object {
        val TRY_RECONNECT_TIMES_INFINITE: Int = -1
    }
}